Lock pyright version

artificial-aidan opened the pull request:
Took a stab at #23. A couple of caveats.

1. `macos-latest` recently deprecated python 3.6, so we removed that. https://github.com/actions/virtual-environments/issues/4060
2. The `version.py` script got a little messy, would take some input on cleaning that up, or another way to do it
3. To test this, I set the `__pyright_version__` 1 behind, which lets the workflow get triggered and a PR created. See here for an automated PR: https://github.com/artificialinc/pyright-python/pull/9

Open to input on the whole process, but wanted to get something started.

artificial-aidan commented:
One other thing.

We specifically grabbed the latest version from the npm registry, as that is where the pyright wrapper pulls it from. Didn't want to trigger off of github releases and have npm be out of sync possibly.

RobertCraigie commented on dev-requirements.txt:
```diff
@@ -1 +1,2 @@
 tox
+semver
```
Updating to a beta version adds type hints which fixes one of the `# type: ignore` comments in `version.py`
```suggestion
semver==3.0.0.dev3
```

RobertCraigie commented on pyright/cli.py:
```diff
@@ -6,6 +6,7 @@
 
 from . import node
 from .utils import env_to_bool
+from . import __pyright_version__
```
I prefer grouped imports
```suggestion
from . import __pyright_version__, node
from .utils import env_to_bool
```

artificial-aidan replied:
Agreed, the last projects i was working in had isort enabled, so i have stopped thinking about grouping them myself.

RobertCraigie replied:
Yeah I should probably enable isort for this project too.

RobertCraigie commented on tests/test_main.py:
```diff
@@ -23,6 +25,34 @@ def test_module_invocation() -> None:
+    output = proc.stdout.decode('utf-8')
+    match = VERSION_REGEX.match(output)
+    assert match is not None
+    assert version.parse(match.group(1)) > version.parse('1.1.223')
```
We should probably check that the version is greater than or equal to the auto-updated pyright version
```suggestion
    assert version.parse(match.group(1)) > version.parse(__pyright_version__)
```

RobertCraigie commented on version.py:
```diff
@@ -0,0 +1,76 @@
+
+
+def set_pyright_ver(ver: str):
+    with FileInput('pyright/__init__.py', inplace=True) as f:  # type: ignore
+        for line in f:  # type: ignore
+            line = re.sub(
+                r'^__pyright_version__\s*=\s*[\'"]([^\'"]*)[\'"]',
+                f"__pyright_version__ = '{ver}'",
+                line.rstrip(),  # type: ignore
+            )  # type: ignore
+            print(line)
+
+
+def bump_pyright_package_ver():
+    with FileInput('pyright/__init__.py', inplace=True) as f:  # type: ignore
+        for line in f:  # type: ignore
+            line = line.rstrip()  # type: ignore
+            current_version = re.search(
+                r'^__version__\s*=\s*[\'"]([^\'"]*)[\'"]', line  # type:ignore
+            )
+            if current_version:
+                line = f"__version__ = '{semver.VersionInfo.parse(current_version.group(1)).bump_patch()}'"  # type:ignore
+            print(line)  # type: ignore
```
These `# type: ignore` comments can be removed by switching from `FileInput` to `fileinput.input()`
```diff
- with FileInput('pyright/__init__.py', inplace=True) as f:
+ with fileinput.input('pyright/__init__.py', inplace=True) as f
```

artificial-aidan replied:
Ah, i knew there had to be a simple way.

RobertCraigie reviewed (changes requested):
Some nitpicks.

Also if you are okay with giving me write access to your fork then I'd be more than happy to just push the changes myself.

RobertCraigie commented:
We also need a mechanism to automatically release the package on PyPi.

Do you have any suggestions?

artificial-aidan commented:
I saw you have the release.py script. I think calling that but let it take in some credentials would work. You would have to add repo credentials that match the ones you use for pypi.

I could give that a shot on our local repo, and local pypi registry.

RobertCraigie commented:
Yeah it's more how do we automatically trigger that script?

We could have another workflow that runs on every commit to main checking if the package version has been changed and making a release then.

artificial-aidan commented:
> Yeah it's more how do we automatically trigger that script?
> 
> We could have another workflow that runs on every commit to main checking if the package version has been changed and making a release then.

Ah. I think you could just run it every time a push to main happens. Pypi will reject it if you try and re-publlish an existing package

RobertCraigie commented:
I think I've found a better solution.

Move all the version updating and setting to it's own file (`pyright/_version.py`) and re-export those variables in `__init__.py`, then we can update the release workflow so that it only runs when `pyright/_version.py` changes.

`_version.py`
```py
__version__ = '0.0.13'
__pyright_version__ = '1.1.223'
```

`__init__.py`
```py
from ._version import (
    __version__ as __version__,
    __pyright_version__ as __pyright_version__,
)
```

You would then need to:
- update the `version.py` script to change `_version.py` instead of `__init__.py`
- update `setup.py` to read the version from `_version.py`

and that should be all you have to change

artificial-aidan commented:
Ok, moved that stuff around, and added a release workflow that will run after the Test workflow on the main branch.

Couple todos, I'm testing it on our in house repo, so there are some things to change before merging it to your branch. (pypi repository url and runner tags)

Where did we land on matching version to pyright version?

artificial-aidan commented:
> ```python
> from ._version import (
>     __version__ as __version__,
>     __pyright_version__ as __pyright_version__,
> )
> ```

Codacy reports the `__version__` import as unused. Should I remove that one?

RobertCraigie commented:
> Codacy reports the __version__ import as unused. Should I remove that one?

No I will just tell codacy to ignore the issue. I honestly do not know why I have it enabled for this repo.

RobertCraigie commented:
> Where did we land on matching version to pyright version?

I think matching versions is a good idea.

RobertCraigie commented:
Let me know when it's ready for review :)

artificial-aidan commented:
Ok. So I think this PR is ready for a final review.

Actions needed:
1. Repository secrets for PyPi login. (PYPI_USER, PYPI_PASS)
2. Merge to main and do a manual workflow run to generate a PR?

Right now I set the version to `0.0.13.post0` and `1.1.223` so it will first publish `0.0.13.post0` and then it should generate a PR for version `1.1.224` of the package when it gets run on the cronjob or manually

RobertCraigie commented on version.py:
```diff
@@ -0,0 +1,77 @@
+            print(line)
+
+
+def bump_pyright_package_ver():
+    with fileinput.input('pyright/_version.py', inplace=True) as f:
+        for line in f:
+            line = line.rstrip()
+            current_version = re.search(r'^__version__\s*=\s*[\'"]([^\'"]*)[\'"]', line)
+            if current_version:
+                line = f"__version__ = '{semver.VersionInfo.parse(current_version.group(1)).bump_patch()}'"
+            print(line)
+
```
This appears to not be used anymore

RobertCraigie commented on dev-requirements.txt:
```diff
@@ -1 +1,4 @@
 tox
+semver==3.0.0.dev3
```
semver is no longer needed as `bump_pyright_package_ver` is not used.

RobertCraigie reviewed (changes requested):
Minor nitpicks, thank you! :)

RobertCraigie commented:
Thanks again!

artificial-aidan commented:
@RobertCraigie I'm pretty sure the combination of `workflow_run` and `paths` isn't working. I thought I had tested it but maybe it isn't supported.

I was trying to reuse workflows, but it might be easier to just copy over the test.yml steps in to the release.yml, and have it run on a push to main only if `pyright/_version.py` changes. Basically just have the test job defined in release.yml and test.yml, and make the `publish_release` job require the test job to run first.

RobertCraigie commented:
@artificial-aidan That is unfortunate. In that case I think I'll just remove `workflow_run` from `release.yml` as all releases will be made through a PR which I will not merge unless tests pass anyway.

Thanks for the heads up.
